src/Model/Resolver/Product/PriceRange.php

```diff
@@ -135,7 +137,53 @@ protected function formatPrice(float $regularPrice, float $finalPrice, StoreInte
                 'value' => $finalPrice,
                 'currency' => $store->getCurrentCurrencyCode()
             ],
-            'discount' => $this->discount->getDiscountByDifference($regularPrice, $finalPrice),
+            'discount' => $discount,
         ];
     }
+
+    /**
+     * Calculates correct discount amount
+     * - Bundle items can contain $regularPrice and $finalFrice from two different
+     * - product instances, thus we are intersted in BE set special price procentage.
+     *
+     * @param Product $product
+     * @param float $regularPrice
+     * @param float $finalPrice
+     * @return array
+     */
+    protected function calculateDiscount(Product $product, float $regularPrice, float $finalPrice) : array
+    {
+        if ($product->getTypeId() != 'bundle') {
+            return $this->discount->getDiscountByDifference($regularPrice, $finalPrice);
+        }
+
+        $specialPrice = $this->getSpecialProductPrice($product);
+        $percentOff = ($specialPrice < 0) ? 0 : 100 - $specialPrice;
+
+        return [
+            'amount_off' => $regularPrice * ($percentOff / 100),
+            'percent_off' => $percentOff
+        ];
+    }
+
+    /**
+     * Gets [active] special price value
+     *
+     * @param Product $product
+     * @return float
+     */
+    protected function getSpecialProductPrice(Product $product): float
+    {
+        $specialPrice = $product->getSpecialPrice();
+        if (!$specialPrice) {
+            return -1;
```

Comment:
If an item does not have a special price, it is not -1, it is `undefined`, and if the function determined that it does not exist, it is better to return `null`. `-1` value is misleading and somewhere deeper inside the application's data flow that would be rising questions.

Reply:
Switched to null